plugin/evm/atomic_trie_height_map_repair.go

```diff
@@ -0,0 +1,133 @@
+// (c) 2020-2021, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package evm
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/coreth/core/types"
+	"github.com/ava-labs/coreth/trie/trienode"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+const (
+	repairDone = math.MaxUint64 // used as a marker for when the height map is repaired
+
+	iterationsPerDelay = 1000                   // after this many iterations, pause for [iterationDelay]
+	iterationDelay     = 100 * time.Millisecond // delay between iterations of the repair loop
+)
+
+func (a *atomicTrie) RepairHeightMap(to uint64, iterationDelay time.Duration) (bool, error) {
+	repairFrom, err := database.GetUInt64(a.metadataDB, heightMapRepairKey)
+	switch {
+	case errors.Is(err, database.ErrNotFound):
+		repairFrom = 0 // height map not repaired yet, start at 0
+	case err != nil:
+		return false, err
+	case repairFrom == repairDone:
+		// height map already repaired, nothing to do
+		return false, nil
+	}
+	return true, a.repairHeightMap(repairFrom, to, iterationDelay)
+}
+
+func (a *atomicTrie) repairHeightMap(from, to uint64, iterationDelay time.Duration) error {
+	// open the atomic trie at the last known root with correct height map
+	// correspondance
+	fromRoot, err := getRoot(a.metadataDB, from)
+	if err != nil {
+		return fmt.Errorf("could not get root at height %d: %w", from, err)
+	}
+	hasher, err := a.OpenTrie(fromRoot)
+	if err != nil {
+		return fmt.Errorf("could not open atomic trie at root %s: %w", fromRoot, err)
+	}
+
+	// hashes values inserted in [hasher], and stores the result in the height
+	// map at [commitHeight]. Additionally, it updates the resume marker and
+	// re-opens [hasher] to respect the trie's no use after commit invariant.
+	var (
+		lastLog = time.Now()
+		logEach = 5 * time.Minute
+	)
+	commitRepairedHeight := func(commitHeight uint64) error {
+		root, nodes := hasher.Commit(false)
+		if nodes != nil {
+			err := a.trieDB.Update(root, types.EmptyRootHash, trienode.NewWithNodeSet(nodes))
+			if err != nil {
+				return err
+			}
+		}
+		err = a.trieDB.Commit(root, false)
```

Comment:
do we need to commit here if nodes are nil?

Reply:
no, updated